ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
                     }
                 }
-
-                // Stop the tracer
-                let sir_trace = Rc::get_mut(&mut self.inner)
-                    .unwrap()
-                    .tracer
-                    .take()
-                    .unwrap()
-                    .0
-                    .stop_tracing()
-                    .unwrap();
-
-                // Start a compilation thread.
-                let mtx = Arc::new(Mutex::new(None));
-                let mtx_cl = Arc::clone(&mtx);
-                thread::spawn(move || {
-                    let compiled = compile_trace::<I>(sir_trace).unwrap();
-                    *mtx_cl.lock().unwrap() = Some(Box::new(compiled));
-                    // FIXME: although we've now put the compiled trace into the mutex, there's no
-                    // guarantee that the Location for which we're compiling will ever be executed
-                    // again. In such a case, the memory has, in essence, leaked.
-                });
-                let new_state = State::phase_compiling(mtx);
-                loc.store(new_state, Ordering::Release);
-
-                return None;
             }
-            PHASE_COMPILING => {
-                // We need to free the memory allocated earlier. To ensure we don't race with
-                // another thread and both try and free the memory, we need to transition to
-                // PHASE_LOCKED so that only this thread will try to free the memory.
-                if loc.compare_and_swap(lp, State::phase_locked(), Ordering::Acquire) != lp {
-                    // We raced with another thread that's probably transitioned this Location to
-                    // PHASE_LOCK.
-                    return None;
+            let hl = unsafe { ls.hot_location() };
+            let hl_ptr = hl as *mut _ as *mut ();
+            match hl {
+                HotLocation::Compiled(tr) => {
+                    // FIXME: we need to refcount the compiled trace to know how many threads are
+                    // currently executing it.
+                    let f = unsafe {
+                        mem::transmute::<_, fn(&mut I) -> *mut RawStopgapInterpreter>(tr.ptr())
+                    };
+                    loc.unlock();
+                    return Some(f);
                 }
-                let mtx = unsafe { lp.ref_data::<CompilingTrace<I>>() };
-                match mtx.try_lock() {
-                    Ok(mut gd) => {
-                        if let Some(tr) = (*gd).take() {
-                            let f = unsafe {
-                                mem::transmute::<_, fn(&mut I) -> *mut RawStopgapInterpreter>(
-                                    tr.ptr(),
-                                )
-                            };
-                            loc.store(State::phase_compiled(tr), Ordering::Release);
-                            return Some(f);
+                HotLocation::Compiling(mtx) => {
+                    let tr = {
```

Comment:
I don't understand what is going on here. Comment?

Reply:
Do the `// Compilation is ongoing.` comments within this `if` not explain things a bit?

Reply:
It appears that compilation can be ongoing in two different ways?

Reply:
Not really: we just have to handle a potential race condition, so we have to do two levels of check (first the mutex; then the `Option`).